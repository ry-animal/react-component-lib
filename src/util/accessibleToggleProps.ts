export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
