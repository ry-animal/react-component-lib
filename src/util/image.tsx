import { ISelectedFile } from "src/components/Forms/FileUpload/ISelectedFile";

export const makeDataUrl = (selectedFile: ISelectedFile) =>
  URL.createObjectURL(selectedFile.file);
