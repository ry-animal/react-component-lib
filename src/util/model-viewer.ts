/*
 * This file is a hacky way of allowing any other component to
 * use the <model-viewer /> tag just by doing:
 *   import "../../../util/model-viewer";
 *
 * The authors of the package do not expose it in a React-friendly
 * way, see: https://github.com/google/model-viewer/issues/1502
 *
 * The component documentation is at https://modelviewer.dev/docs/
 */

import "@google/model-viewer";

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    export interface IntrinsicElements {
      "model-viewer": {
        src?: string;
        alt?: string;
        poster?: string;
        autoplay?: boolean;
        exposure?: number;
        "camera-controls"?: boolean;
        "environment-image"?: string;
        "seamless-poster"?: string;
        loading?: "auto" | "lazy" | "eager";
        reveal?: "auto" | "interaction" | "manual";
      } & React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement>,
        HTMLElement
      >;
    }
  }
}
