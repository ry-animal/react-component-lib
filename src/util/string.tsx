const DEFAULT_LOCALE = "en-US";

export const truncateEthAddress = (str: string): string => {
  if (str.length) {
    const numStartingCharacters = 6;
    const numTrailingCharacters = 4;
    return `${str.substring(0, numStartingCharacters)}...${str.substring(
      str.length - numTrailingCharacters,
      str.length
    )}`;
  }
  return "";
};

export const formatUSDPriceString = (value: number): string => {
  const minimumFractionDigits = 2;
  const maximumFractionDigits = 2;
  return `$${value.toLocaleString(DEFAULT_LOCALE, {
    minimumFractionDigits,
    maximumFractionDigits,
  })}`;
};

export const formatETHPriceString = (priceInEth: number): string => {
  const minimumFractionDigits = 0;
  const maximumFractionDigits = 8;
  const maxDigitsWithDecimal = 8;

  let formatted = `${priceInEth
    .toLocaleString(DEFAULT_LOCALE, {
      minimumFractionDigits,
      maximumFractionDigits,
    })
    .replace(/,/gu, "")
    .substring(0, maxDigitsWithDecimal)}`;

  if (formatted.endsWith(".")) {
    formatted = formatted.slice(0, -1);
  }
  return formatted;
};

const evaluateNumberLessThanOne = (num: number) => {
  if (num === 0) {
    return "0";
  }

  const minDecimalValue = 0.0001;
  if (num < minDecimalValue) {
    return minDecimalValue.toString();
  }
  // To round decimal to 4 digits
  const maxDigits = 10000;
  return `${Math.round(num * maxDigits) / maxDigits}`;
};

// Abbreviate Number https://stackoverflow.com/a/9462382

export const abbreviateNumber = (num: number): string => {
  const lookup = [
    { value: 1, symbol: "" },
    { value: 1e3, symbol: "k" },
    { value: 1e6, symbol: "M" },
    { value: 1e9, symbol: "B" },
    { value: 1e12, symbol: "T" },
    { value: 1e15, symbol: "P" },
    { value: 1e18, symbol: "E" },
  ];

  const rx = /\.0+$|(?<id>\.[0-9]*[1-9])0+$/u;
  const digits = 1;

  const item = lookup
    .slice()
    .reverse()
    .find((x) => num >= x.value);
  return item
    ? (num / item.value).toFixed(digits).replace(rx, "$1") + item.symbol
    : evaluateNumberLessThanOne(num);
};

export const getDomainFromRegex = (url: string) => {
  const matches = url.match(/:\/\/(?<id>.[^/]+)/u);
  // Extract domain (will be null if no match is found)
  return matches && matches[1];
};
