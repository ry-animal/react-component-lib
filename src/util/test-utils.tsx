import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
