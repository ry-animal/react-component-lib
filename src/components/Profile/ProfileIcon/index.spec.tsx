import { ProfileIcon } from ".";
import { render } from "../../../util/test-utils";

describe("<ProfileIcon />", () => {
  describe("rendering", () => {
    test("Default rendering matches snapshot (default size)", () => {
      expect(
        render(<ProfileIcon imgSrc="cat.png" hex={false} />).asFragment()
      ).toMatchSnapshot();
    });

    test("Default rendering matches snapshot (custom size)", () => {
      expect(
        render(
          <ProfileIcon imgSrc="cat.png" size={100} hex={false} />
        ).asFragment()
      ).toMatchSnapshot();
    });
    test("Default rendering matches snapshot (hex)", () => {
      expect(
        render(<ProfileIcon imgSrc="cat.png" hex={true} />).asFragment()
      ).toMatchSnapshot();
    });
    test("Default rendering matches snapshot (hex + custom size)", () => {
      expect(
        render(
          <ProfileIcon imgSrc="cat.png" size={100} hex={true} />
        ).asFragment()
      ).toMatchSnapshot();
    });
    test("Default rendering matches snapshot (borderSize)", () => {
      expect(
        render(
          <ProfileIcon imgSrc="cat.png" size={100} borderSize={10} />
        ).asFragment()
      ).toMatchSnapshot();
    });
  });
  test("Default rendering matches snapshot (borderColor)", () => {
    expect(
      render(
        <ProfileIcon imgSrc="cat.png" size={100} borderColor="#333" />
      ).asFragment()
    ).toMatchSnapshot();
  });
  test("Default rendering matches snapshot (transparentBorder)", () => {
    expect(
      render(
        <ProfileIcon imgSrc="cat.png" size={100} transparentBorder={true} />
      ).asFragment()
    ).toMatchSnapshot();
  });
  test("Default rendering matches snapshot (hex + border)", () => {
    expect(
      render(
        <ProfileIcon
          imgSrc="cat.png"
          size={100}
          hex={true}
          transparentBorder={true}
          borderColor="#333"
          borderSize={10}
        />
      ).asFragment()
    ).toMatchSnapshot();
  });
});
