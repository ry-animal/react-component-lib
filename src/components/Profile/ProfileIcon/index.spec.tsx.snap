// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileIcon /> Default rendering matches snapshot (borderColor) 1`] = `
<DocumentFragment>
  .c1 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 100px;
  width: 100px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  border-radius: 50%;
}

.c0 {
  position: relative;
  height: 104px;
  width: 104px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c0:before {
  content: "";
  background-color: #333;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

<div
    class="c0"
    size="100"
  >
    <div
      class="c1 c2"
      size="100"
    />
  </div>
</DocumentFragment>
`;

exports[`<ProfileIcon /> Default rendering matches snapshot (hex + border) 1`] = `
<DocumentFragment>
  .c0 {
  position: absolute;
}

<div
    class="c0"
  >
    <svg
      fill="none"
      height="0"
      viewBox="0 0 50 50"
      width="0"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <clippath
          clipPathUnits="objectBoundingBox"
          id="hexPath"
          transform="scale(0.02, 0.02)"
        >
          <path
            d="M26.2,0.7l19.2,11.1c0.7,0.4,1.2,1.2,1.2,2.1v22.2c0,0.9-0.5,1.7-1.2,2.1L26.2,49.3c-0.7,0.4-1.7,0.4-2.4,0L4.5,38.2c-0.7-0.4-1.2-1.2-1.2-2.1V13.9c0-0.9,0.5-1.7,1.2-2.1L23.8,0.7C24.5,0.3,25.5,0.3,26.2,0.7z"
          />
        </clippath>
      </defs>
    </svg>
  </div>
  .c0 {
  position: relative;
  height: 120px;
  width: 120px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-clip-path: url(#hexPath);
  clip-path: url(#hexPath);
}

.c0:before {
  content: "";
  background-color: #333;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
  opacity: 50%;
}

.c1 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 100px;
  width: 100px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  -webkit-clip-path: url(#hexPath);
  clip-path: url(#hexPath);
}

<div
    class="c0"
    size="100"
  >
    <div
      class="c1 c2"
      size="100"
    />
  </div>
</DocumentFragment>
`;

exports[`<ProfileIcon /> Default rendering matches snapshot (transparentBorder) 1`] = `
<DocumentFragment>
  .c1 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 100px;
  width: 100px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  border-radius: 50%;
}

.c0 {
  position: relative;
  height: 104px;
  width: 104px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c0:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
  opacity: 50%;
}

<div
    class="c0"
    size="100"
  >
    <div
      class="c1 c2"
      size="100"
    />
  </div>
</DocumentFragment>
`;

exports[`<ProfileIcon /> rendering Default rendering matches snapshot (borderSize) 1`] = `
<DocumentFragment>
  .c1 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 100px;
  width: 100px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  border-radius: 50%;
}

.c0 {
  position: relative;
  height: 120px;
  width: 120px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c0:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

<div
    class="c0"
    size="100"
  >
    <div
      class="c1 c2"
      size="100"
    />
  </div>
</DocumentFragment>
`;

exports[`<ProfileIcon /> rendering Default rendering matches snapshot (custom size) 1`] = `
<DocumentFragment>
  .c1 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 100px;
  width: 100px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  border-radius: 50%;
}

.c0 {
  position: relative;
  height: 104px;
  width: 104px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c0:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

<div
    class="c0"
    size="100"
  >
    <div
      class="c1 c2"
      size="100"
    />
  </div>
</DocumentFragment>
`;

exports[`<ProfileIcon /> rendering Default rendering matches snapshot (default size) 1`] = `
<DocumentFragment>
  .c1 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 40px;
  width: 40px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  border-radius: 50%;
}

.c0 {
  position: relative;
  height: 44px;
  width: 44px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  border-radius: 50%;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c0:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

<div
    class="c0"
    size="40"
  >
    <div
      class="c1 c2"
      size="40"
    />
  </div>
</DocumentFragment>
`;

exports[`<ProfileIcon /> rendering Default rendering matches snapshot (hex + custom size) 1`] = `
<DocumentFragment>
  .c0 {
  position: absolute;
}

<div
    class="c0"
  >
    <svg
      fill="none"
      height="0"
      viewBox="0 0 50 50"
      width="0"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <clippath
          clipPathUnits="objectBoundingBox"
          id="hexPath"
          transform="scale(0.02, 0.02)"
        >
          <path
            d="M26.2,0.7l19.2,11.1c0.7,0.4,1.2,1.2,1.2,2.1v22.2c0,0.9-0.5,1.7-1.2,2.1L26.2,49.3c-0.7,0.4-1.7,0.4-2.4,0L4.5,38.2c-0.7-0.4-1.2-1.2-1.2-2.1V13.9c0-0.9,0.5-1.7,1.2-2.1L23.8,0.7C24.5,0.3,25.5,0.3,26.2,0.7z"
          />
        </clippath>
      </defs>
    </svg>
  </div>
  .c0 {
  position: relative;
  height: 104px;
  width: 104px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-clip-path: url(#hexPath);
  clip-path: url(#hexPath);
}

.c0:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

.c1 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 100px;
  width: 100px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  -webkit-clip-path: url(#hexPath);
  clip-path: url(#hexPath);
}

<div
    class="c0"
    size="100"
  >
    <div
      class="c1 c2"
      size="100"
    />
  </div>
</DocumentFragment>
`;

exports[`<ProfileIcon /> rendering Default rendering matches snapshot (hex) 1`] = `
<DocumentFragment>
  .c0 {
  position: absolute;
}

<div
    class="c0"
  >
    <svg
      fill="none"
      height="0"
      viewBox="0 0 50 50"
      width="0"
      xmlns="http://www.w3.org/2000/svg"
    >
      <defs>
        <clippath
          clipPathUnits="objectBoundingBox"
          id="hexPath"
          transform="scale(0.02, 0.02)"
        >
          <path
            d="M26.2,0.7l19.2,11.1c0.7,0.4,1.2,1.2,1.2,2.1v22.2c0,0.9-0.5,1.7-1.2,2.1L26.2,49.3c-0.7,0.4-1.7,0.4-2.4,0L4.5,38.2c-0.7-0.4-1.2-1.2-1.2-2.1V13.9c0-0.9,0.5-1.7,1.2-2.1L23.8,0.7C24.5,0.3,25.5,0.3,26.2,0.7z"
          />
        </clippath>
      </defs>
    </svg>
  </div>
  .c0 {
  position: relative;
  height: 44px;
  width: 44px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-clip-path: url(#hexPath);
  clip-path: url(#hexPath);
}

.c0:before {
  content: "";
  background-color: white;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  position: absolute;
}

.c1 {
  background-image: url("cat.png");
  background-color: #EEEEEE;
  background-size: cover;
  background-position: center;
  position: relative;
  height: 40px;
  width: 40px;
  overflow: hidden;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c2 {
  -webkit-clip-path: url(#hexPath);
  clip-path: url(#hexPath);
}

<div
    class="c0"
    size="40"
  >
    <div
      class="c1 c2"
      size="40"
    />
  </div>
</DocumentFragment>
`;
